# Laboratorio N°1 Metodos de Optimización

#### Se importan las bibliotecas a ocupar

In [1]:
using JuMP
using Cbc

## Problema 1

Un consultor en finanzas requiere elegir para uno de sus clientes un conjunto de acciones en  las  cuales  invertir. 

Se declaran las 6 variables con la restriccion de  no negatividad. Luego se escriben las restricciones de la cantidad maxima (5000) y minima (40000) a invertir por inversion, tambien se declaran las restricciones de cantidades minima y maxima a invertir en tipos de acciones.

Finalmente se escribe la funcion objetivo, la cual es el retorno de la accion multiplicada por la cantidad de esta. 


In [2]:
lpModel = Model(with_optimizer(Cbc.Optimizer))
@variable(lpModel, x[1:6] >= 0)


@constraint(lpModel, 5000 <= x[1] <= 40000)
@constraint(lpModel, 5000 <= x[2] <= 40000)
@constraint(lpModel, 5000 <= x[3] <= 40000)
@constraint(lpModel, 5000 <= x[4] <= 40000)
@constraint(lpModel, 5000 <= x[5] <= 40000)
@constraint(lpModel, 5000 <= x[6] <= 40000)

@constraint(lpModel, x[1] + x[2] + x[3] <= 30000)
@constraint(lpModel, x[1] + x[3] + x[5] + x[6] >= 50000)
@constraint(lpModel, x[1] + x[2] + x[3] + x[4] + x[5] + x[6] <= 100000)

@objective(lpModel, Max, (5.3/100)x[1] + (6.2/100)x[2] + (5.1/100)x[3] + (4.9/100)x[4] + (6.5/100)x[5] + (3.4/100)x[6])

JuMP.optimize!(lpModel)

print("Los resultados son; $(JuMP.value.(x))\n-----\n")


Los resultados son; [5000.0, 20000.0, 5000.0, 25000.0, 40000.0, 5000.0]
-----Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Dec 31 2018 

command line - Cbc_C_Interface -solve -quit (default strategy 1)
Presolve 3 (-6) rows, 6 (0) columns and 13 (-6) elements
0  Obj 1570 Primal inf 30000 (1) Dual inf 0.313994 (6)
3  Obj 5755
Optimal - objective value 5755
After Postsolve, objective 5755, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 5755 - 3 iterations time 0.002, Presolve 0.00
Total time (CPU seconds):       0.00   (Wallclock seconds):       0.01



### Resultados

En la accion 1 debe invertirse \\$5000, en la accion 2 debe invertirse \\$20000, en la 3 \\$5000, en la 4 \\$25000, en la 5 \\$40000 y en la 6 \\$5000, lo que nos da un retorno de \\$5755.

## Problema 2



In [3]:
lpModel = Model(with_optimizer(Cbc.Optimizer))
@variable(lpModel, x[1:7] >= 0)

@constraint(lpModel, 0.025x[1] + 0.03x[2] >= 0.02(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))
@constraint(lpModel, 0.003x[3] + 0.9x[4] + 0.96x[5] + 0.004x[6] + 0.006x[7] >= 0.004(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))
@constraint(lpModel, 0.013x[1] + 0.008x[2] + 0.04x[5] + 0.012x[6] >= 0.012(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))
@constraint(lpModel, 0.025x[1] + 0.03x[2] <= 0.03(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))
@constraint(lpModel, 0.003x[3] + 0.9x[4] + 0.96x[5] + 0.004x[6] + 0.006x[7] <= 0.006(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))
@constraint(lpModel, 0.013x[1] + 0.008x[2] + 0.004x[5] + 0.012x[6] <= 0.0165(x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7]))

@constraint(lpModel, 0 <= x[1] <= 400)
@constraint(lpModel, 0 <= x[2] <= 300)
@constraint(lpModel, 0 <= x[3] <= 600)
@constraint(lpModel, 0 <= x[4] <= 500)
@constraint(lpModel, 0 <= x[5] <= 200)
@constraint(lpModel, 0 <= x[6] <= 300)
@constraint(lpModel, 0 <= x[7] <= 250)
@constraint(lpModel, x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7] == 500)

@objective(lpModel, Min, 200x[1] + 250x[2] + 150x[3] + 220x[4] + 240x[5] +200x[6] + 165x[7])

JuMP.optimize!(lpModel)

print("Los resultados son; $(JuMP.value.(x))\n-----\n")

Los resultados son; [400.0, 0.0, 39.7763, 0.0, 2.76127, 57.4624, 0.0]
-----Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Dec 31 2018 

command line - Cbc_C_Interface -solve -quit (default strategy 1)
Presolve 5 (-9) rows, 7 (0) columns and 32 (-20) elements
0  Obj 23596.58 Primal inf 487.52598 (2)
5  Obj 98121.636
Optimal - objective value 98121.636
After Postsolve, objective 98121.636, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 98121.63579 - 5 iterations time 0.002, Presolve 0.00
Total time (CPU seconds):       0.00   (Wallclock seconds):       0.00



In [8]:

lpModel = Model(with_optimizer(Cbc.Optimizer))
@variable(lpModel, x[1:18], binary=true)

@constraint(lpModel, x[9] <= x[1])
@constraint(lpModel, x[9] <= x[2])
@constraint(lpModel, x[9] <= x[3])
@constraint(lpModel, x[10] <= x[2])
@constraint(lpModel, x[10] <= x[3])
@constraint(lpModel, x[10] <= x[4])
@constraint(lpModel, x[11] <= x[3])
@constraint(lpModel, x[11] <= x[4])
@constraint(lpModel, x[11] <= x[5])
@constraint(lpModel, x[12] <= x[4])
@constraint(lpModel, x[12] <= x[5])
@constraint(lpModel, x[12] <= x[6])
@constraint(lpModel, x[13] <= x[5])
@constraint(lpModel, x[13] <= x[6])
@constraint(lpModel, x[13] <= x[7])
@constraint(lpModel, x[14] <= x[6])
@constraint(lpModel, x[14] <= x[7])
@constraint(lpModel, x[14] <= x[8])

@constraint(lpModel, x[15] <= x[9])
@constraint(lpModel, x[15] <= x[10])
@constraint(lpModel, x[15] <= x[11])
@constraint(lpModel, x[16] <= x[10])
@constraint(lpModel, x[16] <= x[11])
@constraint(lpModel, x[16] <= x[12])
@constraint(lpModel, x[17] <= x[11])
@constraint(lpModel, x[17] <= x[12])
@constraint(lpModel, x[17] <= x[13])
@constraint(lpModel, x[18] <= x[12])
@constraint(lpModel, x[18] <= x[13])
@constraint(lpModel, x[18] <= x[14])

@objective(lpModel, Max, ((200 - 100)x[1] + (-100)x[2] + (-100)x[3] + (-100)x[4] + (-100)x[5] +
                         (-100)x[6] + (300 -100)x[7] + (-100)x[8] + (-1000)x[9] + (500 -200)x[10] +
                        (-200)x[11] + (200-200)x[12] + (-200)x[13] + (-1000)x[14] + (-1000)x[15] +
                        (-1000)x[16] + (1000-300)x[17] + (1200-1000)x[18])*10000)
                        
JuMP.optimize!(lpModel)
print("Los resultados son; $(JuMP.value.(x))\n-----\n")

Los resultados son; [1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0]
-----
Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Dec 31 2018 

command line - Cbc_C_Interface -solve -quit (default strategy 1)
Continuous objective value is 4e+06 - 0.00 seconds
Cgl0004I processed model has 19 rows, 13 columns (13 integer (13 of which binary)) and 38 elements
Cbc0038I Initial state - 0 integers unsatisfied sum - 0
Cbc0038I Solution found of -4e+06
Cbc0038I Before mini branch and bound, 13 integers at bound fixed and 0 continuous
Cbc0038I Mini branch and bound did not improve solution (0.00 seconds)
Cbc0038I After 0.00 seconds - Feasibility pump exiting with objective of -4e+06 - took 0.00 seconds
Cbc0012I Integer solution of -4000000 found by feasibility pump after 0 iterations and 0 nodes (0.00 seconds)
Cbc0001I Search completed - best objective -4000000, took 0 iterations and 0 nodes (0.00 seconds)
Cbc0035I Maximum depth 0, 0 variables fixe

In [5]:
lpModel = Model(with_optimizer(Cbc.Optimizer))
@variable(lpModel, y[1:7], binary=true)
@variable(lpModel, x[1:7,1:7], binary=true)

@constraint(lpModel, x[1,1] + x[1,2] + x[1,3] + x[1,4] + x[1,5] + x[1,6] + x[1,7] == 1)
@constraint(lpModel, x[2,1] + x[2,2] + x[2,3] + x[2,4] + x[2,5] + x[2,6] + x[2,7] == 1)
@constraint(lpModel, x[3,1] + x[3,2] + x[3,3] + x[3,4] + x[3,5] + x[3,6] + x[3,7] == 1)
@constraint(lpModel, x[4,1] + x[4,2] + x[4,3] + x[4,4] + x[4,5] + x[4,6] + x[4,7] == 1)
@constraint(lpModel, x[5,1] + x[5,2] + x[5,3] + x[5,4] + x[5,5] + x[5,6] + x[5,7] == 1)
@constraint(lpModel, x[6,1] + x[6,2] + x[6,3] + x[6,4] + x[6,5] + x[6,6] + x[6,7] == 1)
@constraint(lpModel, x[7,1] + x[7,2] + x[7,3] + x[7,4] + x[7,5] + x[7,6] + x[7,7] == 1)

@constraint(lpModel, x[1,1] - y[1] <= 0)
@constraint(lpModel, x[1,2] - y[2] <= 0)
@constraint(lpModel, x[1,3] - y[3] <= 0)
@constraint(lpModel, x[1,4] - y[4] <= 0)
@constraint(lpModel, x[1,5] - y[5] <= 0)
@constraint(lpModel, x[1,6] - y[6] <= 0)
@constraint(lpModel, x[1,7] - y[7] <= 0)

@constraint(lpModel, x[2,1] - y[1] <= 0)
@constraint(lpModel, x[2,2] - y[2] <= 0)
@constraint(lpModel, x[2,3] - y[3] <= 0)
@constraint(lpModel, x[2,4] - y[4] <= 0)
@constraint(lpModel, x[2,5] - y[5] <= 0)
@constraint(lpModel, x[2,6] - y[6] <= 0)
@constraint(lpModel, x[2,7] - y[7] <= 0)

@constraint(lpModel, x[3,1] - y[1] <= 0)
@constraint(lpModel, x[3,2] - y[2] <= 0)
@constraint(lpModel, x[3,3] - y[3] <= 0)
@constraint(lpModel, x[3,4] - y[4] <= 0)
@constraint(lpModel, x[3,5] - y[5] <= 0)
@constraint(lpModel, x[3,6] - y[6] <= 0)
@constraint(lpModel, x[3,7] - y[7] <= 0)

@constraint(lpModel, x[4,1] - y[1] <= 0)
@constraint(lpModel, x[4,2] - y[2] <= 0)
@constraint(lpModel, x[4,3] - y[3] <= 0)
@constraint(lpModel, x[4,4] - y[4] <= 0)
@constraint(lpModel, x[4,5] - y[5] <= 0)
@constraint(lpModel, x[4,6] - y[6] <= 0)
@constraint(lpModel, x[4,7] - y[7] <= 0)

@constraint(lpModel, x[5,1] - y[1] <= 0)
@constraint(lpModel, x[5,2] - y[2] <= 0)
@constraint(lpModel, x[5,3] - y[3] <= 0)
@constraint(lpModel, x[5,4] - y[4] <= 0)
@constraint(lpModel, x[5,5] - y[5] <= 0)
@constraint(lpModel, x[5,6] - y[6] <= 0)
@constraint(lpModel, x[5,7] - y[7] <= 0)

@constraint(lpModel, x[6,1] - y[1] <= 0)
@constraint(lpModel, x[6,2] - y[2] <= 0)
@constraint(lpModel, x[6,3] - y[3] <= 0)
@constraint(lpModel, x[6,4] - y[4] <= 0)
@constraint(lpModel, x[6,5] - y[5] <= 0)
@constraint(lpModel, x[6,6] - y[6] <= 0)
@constraint(lpModel, x[6,7] - y[7] <= 0)

@constraint(lpModel, x[7,1] - y[1] <= 0)
@constraint(lpModel, x[7,2] - y[2] <= 0)
@constraint(lpModel, x[7,3] - y[3] <= 0)
@constraint(lpModel, x[7,4] - y[4] <= 0)
@constraint(lpModel, x[7,5] - y[5] <= 0)
@constraint(lpModel, x[7,6] - y[6] <= 0)
@constraint(lpModel, x[7,7] - y[7] <= 0)

@constraint(lpModel, x[1,1] + x[2,1] <= 1)
@constraint(lpModel, x[1,2] + x[2,2] <= 1)
@constraint(lpModel, x[1,3] + x[2,3] <= 1)
@constraint(lpModel, x[1,4] + x[2,4] <= 1)
@constraint(lpModel, x[1,5] + x[2,5] <= 1)
@constraint(lpModel, x[1,6] + x[2,6] <= 1)
@constraint(lpModel, x[1,7] + x[2,7] <= 1)

@constraint(lpModel, x[1,1] + x[3,1] <= 1)
@constraint(lpModel, x[1,2] + x[3,2] <= 1)
@constraint(lpModel, x[1,3] + x[3,3] <= 1)
@constraint(lpModel, x[1,4] + x[3,4] <= 1)
@constraint(lpModel, x[1,5] + x[3,5] <= 1)
@constraint(lpModel, x[1,6] + x[3,6] <= 1)
@constraint(lpModel, x[1,7] + x[3,7] <= 1)

@constraint(lpModel, x[1,1] + x[4,1] <= 1)
@constraint(lpModel, x[1,2] + x[4,2] <= 1)
@constraint(lpModel, x[1,3] + x[4,3] <= 1)
@constraint(lpModel, x[1,4] + x[4,4] <= 1)
@constraint(lpModel, x[1,5] + x[4,5] <= 1)
@constraint(lpModel, x[1,6] + x[4,6] <= 1)
@constraint(lpModel, x[1,7] + x[4,7] <= 1)

@constraint(lpModel, x[1,1] + x[7,1] <= 1)
@constraint(lpModel, x[1,2] + x[7,2] <= 1)
@constraint(lpModel, x[1,3] + x[7,3] <= 1)
@constraint(lpModel, x[1,4] + x[7,4] <= 1)
@constraint(lpModel, x[1,5] + x[7,5] <= 1)
@constraint(lpModel, x[1,6] + x[7,6] <= 1)
@constraint(lpModel, x[1,7] + x[7,7] <= 1)

@constraint(lpModel, x[2,1] + x[3,1] <= 1)
@constraint(lpModel, x[2,2] + x[3,2] <= 1)
@constraint(lpModel, x[2,3] + x[3,3] <= 1)
@constraint(lpModel, x[2,4] + x[3,4] <= 1)
@constraint(lpModel, x[2,5] + x[3,5] <= 1)
@constraint(lpModel, x[2,6] + x[3,6] <= 1)
@constraint(lpModel, x[2,7] + x[3,7] <= 1)

@constraint(lpModel, x[2,1] + x[4,1] <= 1)
@constraint(lpModel, x[2,2] + x[4,2] <= 1)
@constraint(lpModel, x[2,3] + x[4,3] <= 1)
@constraint(lpModel, x[2,4] + x[4,4] <= 1)
@constraint(lpModel, x[2,5] + x[4,5] <= 1)
@constraint(lpModel, x[2,6] + x[4,6] <= 1)
@constraint(lpModel, x[2,7] + x[4,7] <= 1)

@constraint(lpModel, x[2,1] + x[5,1] <= 1)
@constraint(lpModel, x[2,2] + x[5,2] <= 1)
@constraint(lpModel, x[2,3] + x[5,3] <= 1)
@constraint(lpModel, x[2,4] + x[5,4] <= 1)
@constraint(lpModel, x[2,5] + x[5,5] <= 1)
@constraint(lpModel, x[2,6] + x[5,6] <= 1)
@constraint(lpModel, x[2,7] + x[5,7] <= 1)

@constraint(lpModel, x[2,1] + x[7,1] <= 1)
@constraint(lpModel, x[2,2] + x[7,2] <= 1)
@constraint(lpModel, x[2,3] + x[7,3] <= 1)
@constraint(lpModel, x[2,4] + x[7,4] <= 1)
@constraint(lpModel, x[2,5] + x[7,5] <= 1)
@constraint(lpModel, x[2,6] + x[7,6] <= 1)
@constraint(lpModel, x[2,7] + x[7,7] <= 1)

@constraint(lpModel, x[3,1] + x[6,1] <= 1)
@constraint(lpModel, x[3,2] + x[6,2] <= 1)
@constraint(lpModel, x[3,3] + x[6,3] <= 1)
@constraint(lpModel, x[3,4] + x[6,4] <= 1)
@constraint(lpModel, x[3,5] + x[6,5] <= 1)
@constraint(lpModel, x[3,6] + x[6,6] <= 1)
@constraint(lpModel, x[3,7] + x[6,7] <= 1)

@constraint(lpModel, x[3,1] + x[7,1] <= 1)
@constraint(lpModel, x[3,2] + x[7,2] <= 1)
@constraint(lpModel, x[3,3] + x[7,3] <= 1)
@constraint(lpModel, x[3,4] + x[7,4] <= 1)
@constraint(lpModel, x[3,5] + x[7,5] <= 1)
@constraint(lpModel, x[3,6] + x[7,6] <= 1)
@constraint(lpModel, x[3,7] + x[7,7] <= 1)

@constraint(lpModel, x[4,1] + x[5,1] <= 1)
@constraint(lpModel, x[4,2] + x[5,2] <= 1)
@constraint(lpModel, x[4,3] + x[5,3] <= 1)
@constraint(lpModel, x[4,4] + x[5,4] <= 1)
@constraint(lpModel, x[4,5] + x[5,5] <= 1)
@constraint(lpModel, x[4,6] + x[5,6] <= 1)
@constraint(lpModel, x[4,7] + x[5,7] <= 1)

@constraint(lpModel, x[4,1] + x[6,1] <= 1)
@constraint(lpModel, x[4,2] + x[6,2] <= 1)
@constraint(lpModel, x[4,3] + x[6,3] <= 1)
@constraint(lpModel, x[4,4] + x[6,4] <= 1)
@constraint(lpModel, x[4,5] + x[6,5] <= 1)
@constraint(lpModel, x[4,6] + x[6,6] <= 1)
@constraint(lpModel, x[4,7] + x[6,7] <= 1)

@constraint(lpModel, x[5,1] + x[6,1] <= 1)
@constraint(lpModel, x[5,2] + x[6,2] <= 1)
@constraint(lpModel, x[5,3] + x[6,3] <= 1)
@constraint(lpModel, x[5,4] + x[6,4] <= 1)
@constraint(lpModel, x[5,5] + x[6,5] <= 1)
@constraint(lpModel, x[5,6] + x[6,6] <= 1)
@constraint(lpModel, x[5,7] + x[6,7] <= 1)

@constraint(lpModel, x[6,1] + x[7,1] <= 1)
@constraint(lpModel, x[6,2] + x[7,2] <= 1)
@constraint(lpModel, x[6,3] + x[7,3] <= 1)
@constraint(lpModel, x[6,4] + x[7,4] <= 1)
@constraint(lpModel, x[6,5] + x[7,5] <= 1)
@constraint(lpModel, x[6,6] + x[7,6] <= 1)
@constraint(lpModel, x[6,7] + x[7,7] <= 1)

@objective(lpModel, Min, y[1] + y[2] + y[3] + y[4] + y[5] + y[6] + y[7])

JuMP.optimize!(lpModel)
print("Los resultados son Y; $(JuMP.value.(y))\nResults X; $(JuMP.value.(x))\n------")

Welcome to the CBC MILP Solver 
Version: 2.9.9 
Build Date: Dec 31 2018 

command line - Cbc_C_Interface -solve -quit (default strategy 1)
Continuous objective value is 1 - 0.00 seconds
Cgl0003I 0 fixed, 0 tightened bounds, 84 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 97 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 68 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 34 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 14 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 7 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 6 strengthened rows, 0 substitutions
Cgl0003I 0 fixed, 0 tightened bounds, 4 strengthened rows, 0 substitutions
Cgl0004I processed model has 42 rows, 56 columns (56 integer (56 of which binary)) and 196 elements
Cutoff increment increased from 1e-05 to 0.9999
Cbc0038I Initial state - 12 integers unsatisfied sum - 6
Cbc0